src/riak_cs_gc_batch.erl

```diff
@@ -0,0 +1,313 @@
+    State1 = maybe_fetch_first_key(State),
+    NextState = maybe_start_workers(State1),
+    case has_batch_finished(NextState) of
+        true ->
+            {stop, normal, State};
+        _ ->
+            {next_state, waiting_for_workers, NextState}
+    end.
+
+%% @doc This state initiates the deletion of a file from
+%% a set of manifests stored for a particular key in the
+%% garbage collection bucket.
+waiting_for_workers(_Msg, State) ->
+    {next_state, waiting_for_workers, State}.
+
+%% Synchronous events
+
+%% Some race condition?
+prepare(_, _, State) ->
+    {reply, {error, preparing}, prepare, State, 0}.
+
+waiting_for_workers(_Msg, _From, State) ->
+    {reply, ok, waiting_for_workers, State}.
+
+%% @doc there are no all-state events for this fsm
+handle_event({batch_complete, WorkerPid, WorkerState}, StateName, State0) ->
+    %%?debugFmt("~w", [State0]),%% WorkerState#gc_worker_state.batch_count,
+    State = handle_batch_complete(WorkerPid, WorkerState, State0),
+    %%?debugFmt("StateName ~p, ~p ~w", [StateName, has_batch_finished(State), State]),
+    case {has_batch_finished(State), StateName} of
+        {true, _} ->
+            {stop, normal, State};
+        {false, waiting_for_workers} ->
+            try_next_batch(State)
+    end;
+handle_event(_Event, StateName, State) ->
+    {next_state, StateName, State}.
+
+%% @doc Handle synchronous events that should be handled
+%% the same regardless of the current state.
+-spec handle_sync_event(term(), term(), atom(), ?STATE{}) ->
+                               {reply, term(), atom(), ?STATE{}}.
+handle_sync_event(current_state, _From, StateName, State) ->
+    {reply, {StateName, State}, StateName, State};
+handle_sync_event(stop, _From, _StateName, State) ->
+    _ = cancel_batch(State),
+    {stop, cancel, {ok, State}, State};
+handle_sync_event(_Event, _From, StateName, State) ->
+    ok_reply(StateName, State).
+
+handle_info(_Info, StateName, State) ->
+    {next_state, StateName, State}.
+
+%% @doc TODO: log warnings if this fsm is asked to terminate in the
+%% middle of running a gc batch
+terminate(normal, _StateName, State) ->
+    _ = lager:info("Finished garbage collection: "
+                   "~b seconds, ~p batch_count, ~p batch_skips, "
+                   "~p manif_count, ~p block_count\n",
+                   [elapsed(State?STATE.batch_start), State?STATE.batch_count,
+                    State?STATE.batch_skips, State?STATE.manif_count,
+                    State?STATE.block_count]),
+    riak_cs_gc_manager:finished(State);
+terminate(_Reason, _StateName, _State) ->
+    ok.
+
+%% @doc this fsm has no special upgrade process
+code_change(_OldVsn, StateName, State, _Extra) ->
+    {ok, StateName, State}.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+maybe_fetch_first_key(?STATE{batch_start=BatchStart,
```

Comment:
This function fetches keys regardless of any condition. Then, "maybe" is not needed and s/key/keys/.